internal/internal_workflow_client.go

```diff
@@ -155,98 +155,65 @@ func (wc *workflowClient) StartWorkflow(
 	workflowFunc interface{},
 	args ...interface{},
 ) (*WorkflowExecution, error) {
-	workflowID := options.ID
-	if len(workflowID) == 0 {
-		workflowID = uuid.NewRandom().String()
-	}
-
-	if options.TaskList == "" {
-		return nil, errors.New("missing TaskList")
+	startRequest, err := wc.getWorkflowStartRequest(ctx, "StartWorkflow", options, workflowFunc, args...)
+	if err != nil {
+		return nil, err
 	}
 
-	executionTimeout := common.Int32Ceil(options.ExecutionStartToCloseTimeout.Seconds())
-	if executionTimeout <= 0 {
-		return nil, errors.New("missing or invalid ExecutionStartToCloseTimeout")
-	}
+	var response *s.StartWorkflowExecutionResponse
 
-	decisionTaskTimeout := common.Int32Ceil(options.DecisionTaskStartToCloseTimeout.Seconds())
-	if decisionTaskTimeout < 0 {
-		return nil, errors.New("negative DecisionTaskStartToCloseTimeout provided")
-	}
-	if decisionTaskTimeout == 0 {
-		decisionTaskTimeout = defaultDecisionTaskTimeoutInSecs
-	}
+	// Start creating workflow request.
+	err = backoff.Retry(ctx,
+		func() error {
+			tchCtx, cancel, opt := newChannelContext(ctx, wc.featureFlags)
+			defer cancel()
 
-	// Validate type and its arguments.
-	workflowType, input, err := getValidatedWorkflowFunction(workflowFunc, args, wc.dataConverter, wc.registry)
-	if err != nil {
-		return nil, err
-	}
+			var err1 error
+			response, err1 = wc.workflowService.StartWorkflowExecution(tchCtx, startRequest, opt...)
+			return err1
+		}, createDynamicServiceRetryPolicy(ctx), isServiceTransientError)
 
-	memo, err := getWorkflowMemo(options.Memo, wc.dataConverter)
 	if err != nil {
 		return nil, err
 	}
 
-	searchAttr, err := serializeSearchAttributes(options.SearchAttributes)
-	if err != nil {
-		return nil, err
+	if wc.metricsScope != nil {
+		scope := wc.metricsScope.GetTaggedScope(tagTaskList, options.TaskList, tagWorkflowType, startRequest.WorkflowType.GetName())
```

Comment:
```suggestion
		scope := wc.metricsScope.GetTaggedScope(tagTaskList, options.TaskList, tagWorkflowType, *startRequest.WorkflowType.Name)
```
maybe?  both because it's guaranteed to exist (so we might as well enforce it), and because there's a `*startRequest.WorkflowId` below (also guaranteed to exist).

Reply:
makes sense.